kafka/src/main/java/org/apache/apex/malhar/kafka/AbstractKafkaInputOperator.java

```diff
@@ -0,0 +1,512 @@
+ * 5. Multi-cluster support, one operator can consume data from more than one kafka clusters
+ * 6. Multi-topic support, one operator can subscribe multiple topics
+ * 7. Throughput control support, you can throttle number of tuple for each streaming window
+ */
+public abstract class AbstractKafkaInputOperator<K, V> implements InputOperator, Operator.ActivationListener<Context.OperatorContext>, Operator.CheckpointListener, Partitioner<AbstractKafkaInputOperator>, StatsListener, OffsetCommitCallback
+{
+
+  private static final Logger logger = LoggerFactory.getLogger(AbstractKafkaInputOperator.class);
+
+  public long getMetricsRefreshInterval()
+  {
+    return metricsRefreshInterval;
+  }
+
+  public void setMetricsRefreshInterval(long metricsRefreshInterval)
+  {
+    this.metricsRefreshInterval = metricsRefreshInterval;
+  }
+
+  public enum InitialOffset
+  {
+    EARLIEST, // consume from beginning of the partition every time when application restart
+    LATEST, // consume from latest of the partition every time when application restart
+    APPLICATION_OR_EARLIEST, // consume from committed position from last run or earliest if there is no committed offset(s)
+    APPLICATION_OR_LATEST // consume from committed position from last run or latest if there is no committed offset(s)
+  }
+  /**
+   *  Same setting as bootstrap.servers property to KafkaConsumer
+   *  refer to http://kafka.apache.org/documentation.html#newconsumerconfigs
+   *  To support multi cluster, you can have multiple bootstrap.servers separated by ";"
+   */
+  @NotNull
+  private String[] clusters;
+
+  /**
+   * The topics the operator consumes
+   */
+  @NotNull
+  private String[] topics;
+
+  /**
+   * Wrapper consumer object
+   * It wraps KafkaConsumer, maintains consumer thread and store messages in a queue
+   */
+  private KafkaConsumerWrapper consumerWrapper = new KafkaConsumerWrapper();
+
+  /**
+   * Assignment for each operator instance
+   */
+  private Set<AbstractKafkaPartitioner.PartitionMeta> assignment;
+
+  /**
+   *  offset track for checkpoint
+   */
+  private Map<AbstractKafkaPartitioner.PartitionMeta, Long> offsetTrack = new HashMap<>();
+
+  /**
+   * store offsets with window id, only keep offsets with windows that have not been committed
+   */
+  private transient List<Pair<Long, Map<AbstractKafkaPartitioner.PartitionMeta, Long>>> offsetHistory = new LinkedList<>();
+
+  /**
+   * initial partition count
+   * only used with PartitionStrategy.ONE_TO_MANY
+   */
+  private int initialPartitionCount = 1;
+
+  // Minimal interval between 2 (re)partition actions
+  private long repartitionInterval = 30000L;
+
+  // Minimal interval between checking collected stats and decide whether it needs to repartition or not.
+  // And minimal interval between 2 offset updates
+  private long repartitionCheckInterval = 5000L;
+
+  // class name of the key deserializer
+  protected String keyDeserializer;
```

Comment:
Why String, why not objects? However, given that we convert in emitTuples, we probably don't need these at all.